File: examples/ecommerce-ts/handlers/list_collections.ts
import type { RouteRequest, RouteResult, PrefetchMessage, PrefetchDb, HandleContext, RenderContext } from "focus";
import { esc } from "focus";

// [route] .list_collections
// match GET /collections
export function route(_req: RouteRequest): RouteResult | null {
  return {};
}

// [prefetch] .list_collections
export async function prefetch(_msg: PrefetchMessage, db: PrefetchDb) {
  const collections = await db.queryAll("SELECT id, name, active FROM collections WHERE active = 1 ORDER BY id LIMIT ?1", 50);
  return { collections };
}

// [handle] .list_collections
export function handle(_ctx: HandleContext): string {
  return "ok";
}

// [render] .list_collections
export function render(ctx: RenderContext): string {
  const items = ctx.prefetched.collections as unknown[] | null;
  if (!items || items.length === 0) return "<div>No collections</div>";
  return items.map((c: any) => `<div class="card">${esc(c.name)}</div>`).join("");
}

File: examples/ecommerce-py/handlers/list_collections.py
from tiger_web import esc


def route(req):
    return {"operation": "list_collections", "id": "0" * 32}


def prefetch(msg, db):
    collections = db.query_all(
        "SELECT id, name FROM collections ORDER BY id LIMIT ?1", 50
    )
    return {"collections": collections}


def handle(ctx, db):
    return "ok"


def render(ctx, db=None):
    collections = ctx["prefetched"].get("collections") or []
    if not collections:
        return '<div class="meta">No collections</div>'

    return "".join(
        f'<div class="card"><strong>{esc(c["name"])}</strong>'
        f'<div class="meta">{c["id"]}</div></div>'
        for c in collections
    )
